src/ucp/dt/dt_generic.c

```diff
@@ -0,0 +1,42 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "dt_generic.h"
+
+#include <ucs/debug/memtrack.h>
+
+
+ucs_status_t ucp_dt_create_generic(ucp_generic_dt_ops_t *ops, void *context,
+                                   ucp_datatype_t *datatype_p)
+{
+    ucp_dt_generic_t *dt;
+
+    dt = ucs_memalign(UCS_BIT(UCP_DATATYPE_SHIFT), sizeof(*dt), "generic_dt");
+    if (dt == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    dt->ops      = ops;
+    dt->context  = context;
+    *datatype_p = ((uintptr_t)dt) | UCP_DATATYPE_GENERIC;
```

Comment:
I know that I proposed this (caching flag on bits not used by pointer), but we have to be careful with supporting of this across different architectures. x86_64 is safe. What about POWER and ARM. Probably the offset for unused bits that has to be #ifdefed by platform.